### PR TITLE
Sc 12283/db validation

### PR DESCRIPTION
**Pull request by daniellemaxwell**
### Scope of changes

Adds validation method to the db package for the member, project, and topic models that will be used when creating or updating these resources. The `members` and `projects` db required two validation methods as they have 2 create methods and one requires a `TenantID` while another doesn't.  

Minor updates to server-side handlers and tests were required.

Note: Validation may be added to the tenant model in the future.

Fixes SC-12283

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [x] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Describe how reviewers can test this change to be sure that it works correctly. Add a checklist if possible.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.

### CODE REVIEW

File: pkg/tenant/db/members.go [hunk 1]
```diff
@@ -52,11 +53,73 @@ func (m *Member) UnmarshalValue(data []byte) error {
...
+	// TODO: Use crypto rand and monotonic entropy with ulid.New
+
+	// Validate tenant project data.
+	if err = member.ValidateWithID(); err != nil {
+		return err
+	}
+
+	if member.ID.Compare(ulid.ULID{}) == 0 {
+		member.ID = ulid.Make()
+	}
```
Comment (pdeziel) [inline pkg/tenant/db/members.go:102]: Or maybe we could instead move this check above the validation check to ensure that the member always has an ID before it's validated?

File: pkg/tenant/db/members_test.go [hunk 2]
```diff
@@ -43,10 +46,47 @@ func TestMemberModel(t *testing.T) {
...
+
+		return &pb.PutReply{
+			Success: true,
+		}, nil
+	}
+
+	err = db.CreateMember(ctx, member)
+	require.NoError(err, "could not create member")
+
+	require.NotEqual("", member.ID, "expected non-zero ulid to be populated")
```
Comment (pdeziel) [inline pkg/tenant/db/members_test.go:75]: I think you can also use `require.NotEmpty` which is slightly cleaner.

**Review by pdeziel — COMMENTED** [review]
I think this is a good start! I did have some suggestions in the comments about how we can reduce some of the repeated code, which should make this more maintainable in the future.

File: pkg/tenant/db/members.go [hunk 3]
```diff
@@ -52,11 +53,73 @@ func (m *Member) UnmarshalValue(data []byte) error {
 	return msgpack.Unmarshal(data, m)
 }
 
-// CreateMember adds a new Member to the database.
+func (m *Member) Validate() error {
+	memberName := m.Name
+
+	if strings.ContainsAny(string(memberName[0]), "0123456789") {
```
[the remark this answers is not in the document]
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go:63]: It did cause a panic. An additional check has been added to confirm if a name exists.

File: pkg/tenant/db/members.go [hunk 4]
```diff
@@ -52,11 +53,73 @@ func (m *Member) UnmarshalValue(data []byte) error {
...
+		return ErrNumberFirstCharacter
+	}
+
+	if strings.ContainsAny(memberName, " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~") {
+		return ErrSpecialCharacters
+	}
```
[the remark this answers is not in the document]
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go:69]: Since the methods have been added to each model, I decided to keep doing that for now. Also, the models could change. 

The link didn't work for me and I wasn't able to find exactly what was referenced. Could you try resending?

File: pkg/tenant/db/members.go (hunk 1 above)
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go:108]: In the case of `CreateTenantMember` and `CreateTenantProject`, I think we should check if the `TenantID` exists before checking if a `MemberID` or `ProjectID` exists.

File: pkg/tenant/db/members.go [hunk 5]
```diff
@@ -52,11 +53,73 @@ func (m *Member) UnmarshalValue(data []byte) error {
...
+
+	return nil
+}
+
+// Validates data in the member model when a TenantID is required.
+// Ex. CreateTenantProject and UpdateProject
+func (m *Member) ValidateWithID() error {
```
[the remark this answers is not in the document]
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go]: We want to have a `TenantID` in the models, but this then causes errors with the server-side `CreateMember` and `CreateProject` handlers. I created a `ValidateID` method that only checks for the ID to eliminate the redundant code. Although, this might not exactly be the most optimal solution at the moment.

[the remark this answers is not in the document]
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go]: @pdeziel For now validation will only be applied to `TenantCreateMember`, `TenantCreateProject`, and `ProjectTopicCreate`.

File: pkg/tenant/db/members.go (hunk 1 above)
Comment (daniellemaxwell) [reply pkg/tenant/db/members.go:85]: The `MemberID` check has been moved up before the validation check.

File: pkg/tenant/db/members.go [hunk 6]
```diff
@@ -52,7 +53,52 @@ func (m *Member) UnmarshalValue(data []byte) error {
...
+func (m *Member) Validate() error {
+	if m.TenantID.Compare(ulid.ULID{}) == 0 {
+		return ErrMissingTenantID
+	}
+
+	memberName := m.Name
+
+	if memberName == "" {
+		return ErrMissingMemberName
+	}
```
Comment (pdeziel) [inline pkg/tenant/db/members.go:65]: Nice! Let's add a test case for this and to the other models so that we protect ourselves against regressions when we make future changes to the validation code.

**Review by pdeziel — APPROVED** [review]
Thank you for making the changes; this is so much clearer to me as a reader of the code. I have one small suggestion to add a quick test case to make sure the empty string check is working and another suggestion to add a follow on refactoring story but once those are complete feel free to merge!

File: pkg/tenant/db/members.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdeziel) [reply pkg/tenant/db/members.go]: Can you create a follow up story to focus on refactoring the name checks in a way that reduces the repeated code between the different files? I can add some details on the regex there.

I do think it's a good idea to have different validation methods for the different models (to reduce coupling between them), but we might be able to standardize the input checking a bit.
